host/frontend/host-orchestrator/devices.go

```diff
@@ -0,0 +1,135 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+//     Unless required by applicable law or agreed to in writing, software
+//     distributed under the License is distributed on an "AS IS" BASIS,
+//     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//     See the License for the specific language governing permissions and
+//     limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http/httputil"
+	"net/url"
+	"sync"
+)
+
+type Device struct {
+	info interface{}
+	ws   *JsonWs
+	// Reverse proxy to the client files
+	Proxy *httputil.ReverseProxy
+	// Synchronizes access to the client list and the client count
+	clientsMtx sync.Mutex
+	clients    map[int]Client
+	// Historical client count, it doesn't decrease when clients unregister. It's used to generate client ids.
+	clientCount int
+}
+
+func NewDevice(ws *JsonWs, port int, info interface{}) *Device {
+	url, err := url.Parse(fmt.Sprintf("http://127.0.0.1:%d", port))
+	if err != nil {
+		// This should not happen
+		log.Fatal("Unable to parse hard coded url: ", err)
+		return nil
+	}
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	return &Device{ws: ws, Proxy: proxy, info: info, clients: make(map[int]Client), clientCount: 0}
+}
+
+// Sends a message to the device
+func (d *Device) Send(msg interface{}) bool {
+	return d.ws.Send(msg)
+}
+
+func (d *Device) Register(c Client) int {
+	d.clientsMtx.Lock()
+	defer d.clientsMtx.Unlock()
+	d.clientCount += 1
+	d.clients[d.clientCount] = c
+	return d.clientCount
+}
+
+func (d *Device) Unregister(id int) {
+	d.clientsMtx.Lock()
+	defer d.clientsMtx.Unlock()
+	delete(d.clients, id)
+}
+
+// Notify the clients that the device has disconnected
+func (d *Device) DisconnectClients() {
+	d.clientsMtx.Lock()
+	defer d.clientsMtx.Unlock()
+	for _, client := range d.clients {
+		client.OnDeviceDisconnected()
+	}
+}
+
+// Send a message to the client
+func (d *Device) ToClient(id int, msg interface{}) bool {
+	d.clientsMtx.Lock()
+	client, ok := d.clients[id]
+	d.clientsMtx.Unlock()
+	if !ok {
+		return false
+	}
+	return client.Send(msg)
+}
+
+// Keeps track of the registered devices.
+type DevicePool struct {
+	devicesMtx sync.Mutex
+	devices    map[string]*Device
+}
+
+func NewDevicePool() *DevicePool {
+	return &DevicePool{
+		devices: make(map[string]*Device),
+	}
+}
+
+// Register a new device, returns false if a device with that id already exists
+func (p *DevicePool) Register(d *Device, id string) bool {
+	p.devicesMtx.Lock()
+	defer p.devicesMtx.Unlock()
+	_, ok := p.devices[id]
+	if ok {
+		return false
+	}
+	p.devices[id] = d
+	return true
+}
+
+func (p *DevicePool) Unregister(id string) {
+	p.devicesMtx.Lock()
+	defer p.devicesMtx.Unlock()
+	if d, ok := p.devices[id]; ok {
+		d.DisconnectClients()
```

Comment:
clients have no access to the device pool.
The device pool is only used to keep track of devices by id, it can be accessed by a thread that handles client requests but the actual client won't need a reference to it because it cares about only one device.
Similarly devices don't need to reference the pool either, so the client won't end up calling to the pool indirectly